Documentation/proposals/202304-silences.md

```diff
@@ -0,0 +1,112 @@
+
+By Adding support for `Silence` CRD in the prometheus-operator, this will make users more flexible in terms of choosing the tool
+to deploy their silences and will free others from managing a standalone component within the stack.
+
+## Pitfalls of the current solution
+
+Using Alertmanager API Directly comes with drawbacks:
+
+* Teams have to build an automation to add silences in a centralized manner
+* There is no input validation, which can lead to an invalid silence configuration
+* Teams need to manage expiration on their own
+* If a silence was deleted by accident, it will be permanent.
+
+# Goals
+
+* Provide a way for users to manage Alertmanager silences with the Kubernetes Way.
+* Make it easier to manage Alertmanager silences via centralised repo with e.g. ArgoCD, Flux...
+* Provide a way to manage silence expiration with a silence CR.
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have an interface in defining silences exposed to developers.
+* Users who want to manage silences the same way as for services running within the Kubernetes cluster
+* Users who want a supported Kubernetes way of silences outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal doesn't aim to remove the CR after the expiration date has reached
+* Refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+Creating a new cluster-scope Silence CRD that will act as an interface by adding silences via the Alertmanager API.
+
+A typical `Silence` resource could look like the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: Silence
+metadata:
+  name: my-silence
+  labels:
+    test: value
+spec:
+  selector:
+    app: test-alertmanager
+  expiresAt: "2023-06-01"
+  matchers:
+    - name: namespace
+      value: test-ns
+      isRegex: false
+```
+
+The above resource will result in creating a silence in the Alertmanager with label selector,
+`app: test-alertmanager`.
+
+* `selector` used to select which Kubernetes Service to use as an Alertmanager API.
+* `expiresAt` to define the expiration of the silence.
+* `matchers` field corresponds to the Alertmanager silence matchers each of which consists of:
+  - `name` - name of tag on an alert to match
+  - `value` - fixed string or expression to match against the value of the tag named by name above on an alert
+  - `isRegex` - a boolean specifying whether to treat value as a regex (=~) or a fixed string (=)
+  - `isEqual` - a boolean specifying whether to use equal signs (= or =~) or to negate the matcher (!= or !~)
+
+This example doesn't list all the fields that are offered by prometheus. The implementation of all the fields will be
+done in an iterative process and as such, the expectation is not for all of them to be implemented in the first version.
```

Comment:
I might be wrong, but the missing fields are just "Created By" and "Comment", aren't they? Which doesn't seem to offer a big challenge to implement when compared to expiration handling.

Couldn't we just add all fields all at once?

Reply:
we can add them as annotation, or you prefer fields under the spec?

Reply:
Any thoughts here?